dwave/experimental/automorphism/automorphism_generation.py

```diff
@@ -0,0 +1,655 @@
+        and Algorithms 1 and 2 from Berkholz, C. (2016). Tight lower and upper
+        bounds for the complexity of canonical colour refinement.
+
+        A stack is initialized to contain each block in the initial partition.
+        An invariant h equal to the length of the intersection of the neighbourhood
+        of a node with blocks popped from the stack is used to iteratively refine
+        the initial partition. If a block successfully refines the partition, the
+        new sub-blocks are added to the stack. Hopcroft's trick enables us to discard
+        one of these blocks — a more intentional way of determining which block to
+        discard may yield better performance.
+
+        Keeping track of the neighbourhood of each block popped from the stack
+        enables the refinement algorithm to be O(m log(n)) as opposed to O(n**2 log(n)),
+        where m and n are edges and vertices, respectively, which significantly
+        improves the algorithm for sparse graphs.
+
+        Args:
+            partition: Partition represented as a list of vertex index sets.
+        """
+        neighbours = self._neighbours
+        vertex_block_index = self._vertex_block_index
+
+        remaining_vertices = set(self._nodes)
+        blocks_stack = deque(partition)
+        while blocks_stack:
+            current_block = blocks_stack.pop()
+            if current_block <= remaining_vertices:
+                remaining_vertices -= current_block
+                touched_blocks = set()
+                for v in current_block:
+                    for w in neighbours[v]:
+                        touched_blocks.add(vertex_block_index[w])
+                vertex_block_index = self._vertex_block_index
+
+                for block_index in sorted(touched_blocks, reverse=True):
+                    count_to_vertices = defaultdict(set)
+                    for u in partition[block_index]:
+                        count = len(current_block & neighbours[u])
+                        count_to_vertices[count].add(u)
+                    num_new_blocks = len(count_to_vertices)
+                    if num_new_blocks > 1:
+                        len_partition = len(partition)
+                        for _ in range(num_new_blocks - 1):
+                            partition.append(set())
+                        for count in range(len_partition - 1, block_index, -1):
+                            partition[num_new_blocks - 1 + count] = partition[count]
+                        new_blocks = []
+                        offset = 0
+                        for count_key in sorted(count_to_vertices):
+                            partition[block_index + offset] = count_to_vertices[count_key]
+                            remaining_vertices.update(count_to_vertices[count_key])
+                            new_blocks.append(count_to_vertices[count_key])
+                            offset += 1
```

Comment:
slightly simpler with `enumerate`:
```suggestion
                        for offset, count_key in enumerate(sorted(count_to_vertices)):
                            partition[block_index + offset] = count_to_vertices[count_key]
                            remaining_vertices.update(count_to_vertices[count_key])
                            new_blocks.append(count_to_vertices[count_key])
```